Address book and controller tests

pvpetit opened the pull request:
AddressBook and Controller Tests.

pvpetit committed:
Nothing Changed.

pvpetit committed:
Address Book and AddressBookController Tests

pvpetit committed:
Extended Coverage.
Added Person Test with new conditions.

ZGreening commented on src/AddressBook.java:
```diff
@@ -1,19 +1,33 @@
+
 
 public class AddressBook extends AbstractTableModel {
-    private static final long serialVersionUID = 1L;
```
This variable is expected in classes that implement serializable. Can you please add it back?

ZGreening replied:
VSCode highlights this as an warning if it is not there

pvpetit replied:
Added the warning.

ZGreening commented on src/AddressBook.java:
```diff
@@ -1,19 +1,33 @@
-    private static final long serialVersionUID = 1L;
     private List<Person> persons = new ArrayList<>();
 
+    public List<Person> getPersonsList(){
```
Do we need to add this?

pvpetit replied:
Its for the testcases because there's no direct access to Address Book's person List.

ZGreening replied:
Normally code isn't supposed to have direct access to objects like array lists. there is a getPersons() function right below which returns the same data in array format. The reason this way is preferred is the original data cannot be changed or replaced, it is preserved safe.

ZGreening replied:
Can you try and rewrite the code you have that depends on this to just use the array form in the existing getPersons()???

pvpetit replied:
Ok I used getPersons() instead.

ZGreening commented on src/AddressBook.java:
```diff
@@ -1,19 +1,33 @@
         int newIndex = persons.size();
         persons.add(p);
-        fireTableRowsInserted(newIndex, newIndex);
+        System.out.println("new index: "+newIndex);
```
New index isn't really being used with your change here. The insert to 0,0 is maybe fine, the old implementation just inserted a person to the last index location instead. Did you run this? Because I am not 100% sure it works with 0,0

pvpetit replied:
It's how the addressBook original code is. I just added the println which doesn't need to be there anymore.

ZGreening replied:
But isn't the original code inserting to (newIndex,newIndex) instead of (0,0)?

pvpetit replied:
(newIndex,newIndex) is what is the code is currently. I don't see 0,0

ZGreening replied:
It's in your code, I accidentally didn't highlight it in GitHub when making the comment.

ZGreening commented on src/AddressBook.java:
```diff
@@ -56,6 +70,9 @@ public int getColumnCount() {
     }
 
+
+    //So I blieve program autimcally populate itself ussing persons ArrayList.
```
The table uses a listener, kinda like a signal (or trap), to update.

ZGreening commented on src/Person.java:
```diff
@@ -1,10 +1,11 @@
 public class Person {
 
-    public static final String[] fields = { "Last Name", "First Name", "Address", "City", "State", "ZIP", "Phone" };
+    public static final String[] fields = { "Last Name", "First Name", "Address", "City", "State", "ZIP", "Phone", };
```
Doesn't the last comma on the list prevent this from compiling?

pvpetit replied:
Thats strange it hasn't given me a error, but it shouldn't be there.

ZGreening commented on src/Person.java:
```diff
@@ -14,12 +15,43 @@
         if (lastName == null || lastName.isEmpty())
             throw new IllegalArgumentException("Last name cannot be empty");
+        //Check if first name passes regex
+        if (checkNameRegex(firstName,"\\S{2,}")== false){
```
I made some changes to the person class. I don't know if yours or mine is better per se. Do you want to have a look at my changes?

ZGreening replied:
Never mind, I never did. I like these changes, I guess we'll go with 9 digit phone numbers, I wrote something below about it, but 9 is probably best.

ZGreening replied:
I was mistaken the other night. It is 10 digits, please edit this

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+  @AfterEach
+  void tearDown() {
+    //model = null;
+    personListTest = null;
```
Honestly, these are unnecessary, I don't know why the professor did that in hers. Java's garbage collection takes care of it all.

pvpetit replied:
Probably but she put in the powerpoint so I'm assuming she expects to be there.

ZGreening replied:
She did, but it is unnecessary. Some programmers think being explicit in commands as much as possible is preferred, others (like me) prefer as little code as possible. Since it is not necessary my natural preference is to delete it... It's fine to leave it though. I don't really care. I was just pointing that out

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+
+    controllerTest.add(test_Person);
+
+    System.out.println(controllerTest.get(0).toString());
```
Usually try not to have print statements in finished code, but it probably doesn't matter for now

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+    controllerTest.add(test_Person);
+
+    System.out.println(controllerTest.get(0).toString());
+    assertEquals(controllerTest.get(0),test_Person);
```
isn't this a pointer comparison. It is probably better to compare first name or something. Unless the Person class implements an overriden "equals" function for comparisons

pvpetit replied:
assertEquals(controllerTest.get(0).getFirstName(),"Jane");
This assert statement checks for the first name. I wasn't sure the best way for the test to check if the object was created successfully.

ZGreening replied:
That could work. You are just testing whether or not the person was added, so putting in a predetermined value like that should work. FYI, just so you know, I discovered that assertEquals() has a specific order for which spot should be the part that is tested. I believe it is "assertEquals(expected, testedCode)" not "assertEquals(testedCode,expected)". It doesn't affect the outcome, but If a test fails the print statements it returns provide the reverse information, Your code would say "Expected "Joe" but received "Jane" if you were to change the name of the object in the first argument.

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+    //excetionRule.expect(IndexOutOfBoundsException.class);
+
+
+    Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
```
Maybe import static Assertions.* ?

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+
+
+
+    //controllerTest.remove(0);
```
Just a style comment, There is a lot of extra white space lines and commented out code

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+//
+//    verify(addressBookMock).clear();
+
+  }
```
Does the controller contain anything before the "clear" function is called? Maybe a better test first adds a person to the list, tests they exist, clears, then tests they no longer exist. Also, "assertEquals(something, true)" would be the same as "assertTrue(something)".

pvpetit replied:
Good idea, I edited the test case.

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+  public void openFile(File file)  {
+//    new FileSystem().readFile(addressBook, file);
+//    addressBook.fireTableDataChanged();
+  }
```
Test?

pvpetit replied:
Any Idea how to test this? Should I try using the temporary Folder?

ZGreening replied:
What I wanted to do was run tests like this sequentially. Like first test save file, then check read file. Unfortunately, JUnit discourages this, so what I did for tests like these was copy the same code from the save file one (removing the assertions), then try to open that saved file, and write assertions to check the file was opened correctly. I think I have similar code written in my GUI one that might help you

pvpetit replied:
I used the code from your File System test. I believe I have a solid test case for this now.

ZGreening commented on tests/AddressBookControllerTest.java:
```diff
@@ -0,0 +1,167 @@
+  //Figure this out later
+  public void save(File file)  {
+    //new FileSystem().saveFile(addressBook, file);
+  }
```
Missing test code, probably to test, just call the controller's save function giving it the proper parameters, then test that the file was saved. I think that should give you 100% code coverage for this function, right? I had to do a bit of inventing to call the lambda functions in the GUI class. When you review my code, have a look at how I handled verifying some of the tests were successful

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+
+import java.util.ArrayList;
```
Extra white space

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+
+
+
```
Same as above

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+    Person[] test_PersonArray = new Person[personList.size()];
+    personList.toArray(test_PersonArray);
+
+    test_AddressBook.getPersons();
```
This is a 'get' call, but it's output is not assigned to anything

pvpetit replied:
Fixed.

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+    //Given an index delete from arrayList
+    test_AddressBook.add(test_Person);
+    test_AddressBook.remove(0);
+    assertEquals(test_AddressBook.getPersonsList().isEmpty(),true);//Confirm delete
```
assertTrue instead of assertEquals

pvpetit replied:
Fixed.

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  @Test
+  void get() {
```
The function name could be better named, 'get' what?

ZGreening replied:
testAddressBookGetPerson() ???

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+  @Test
+  void clearWithEmptyList() {
+    test_AddressBook.clear();
+    assertEquals(test_AddressBook.getPersonsList().isEmpty(),true);
```
assertTrue

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+    assertEquals(test_AddressBook.getColumnCount(),7);
+
+
+  }
```
These test are very good, short sweet and to the point.

ZGreening replied:
a bit of extra white space

ZGreening commented on tests/AddressBookTest.java:
```diff
@@ -0,0 +1,178 @@
+
+    test_AddressBook.add(test_Person);
+    System.out.println(test_AddressBook.getColumnName(1));
+  }
```
This function is missing an assertion

ZGreening commented on tests/PersonTest.java:
```diff
@@ -1,75 +1,329 @@
+  void test1(){
+    test_Person.getCity();
+  }
+
```
There is no "assert" statement to test anything here. Can we delete this test?

ZGreening commented on tests/PersonTest.java:
```diff
@@ -1,75 +1,329 @@
+  void personPhoneNumberMustbe9NumbersNegativeCase(){
+    assertThrows(IllegalArgumentException.class, () -> new Person("John","Doe","123 Fake Street","Fort Myers","FL","33901","0123456789" ));
+    //assertNotNull(new Person("John","Doe","123 Fake Street","Fort Myers","FLR","33901","123456789"));
+  }
```
US phone numbers are 10 digits, +1 (239) 123-4567 counting the country and area code. I created all my tests with 10 digits in mind. Do you want to stay with 9 instead? I can change mine easy enough if so. Maybe 9 is better

ZGreening replied:
Commented above, let's go with what you have, 9 digits.

ZGreening replied:
Oh, wow, I can't believe I mis counted that yesterday. It is 11 digits with country code and 10 with just the area code. I must have been pretty tired at the time. I didn't get much sleep over the whole weekend. I actually had changed all my files to 9 digits before I realized the mistake! Anyway, please correct your tests to use 10 digits.

pvpetit replied:
Changed the regex to accept only 10 digits.

ZGreening commented on tests/PersonTest.java:
```diff
@@ -1,75 +1,329 @@
-import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
 
-import org.junit.jupiter.api.BeforeEach;
+import java.util.Arrays;
+import java.util.List;
```
I guess I am fine to merge this classes code, but this is what Nathalie was supposed to be working on. Most likely it will change at least a bit

ZGreening reviewed (changes requested):
A few minor things, could you please look through my notes and clean the code up a bit. NOTE: Just commit your fixes to the same location, the pull request is updated automatically.

ZGreening commented on src/AddressBook.java:
```diff
@@ -1,19 +1,33 @@
     public void add(Person p) {
         int newIndex = persons.size();
         persons.add(p);
-        fireTableRowsInserted(newIndex, newIndex);
```
It shows newIndex being removed here

pvpetit replied:
Ok I changed it back.

ZGreening commented on src/AddressBook.java:
```diff
@@ -1,19 +1,33 @@
         persons.add(p);
-        fireTableRowsInserted(newIndex, newIndex);
+        System.out.println("new index: "+newIndex);
+        fireTableRowsInserted(0, 0);
```
And 0,0 being added here

pvpetit committed:
Fixed Address Book and Controller Test Cases.
Added Coverage.
Person Class changed so regex for first name last name,zip,state, and
phonenumber(10)

pvpetit commented:
> A few minor things, could you please look through my notes and clean the code up a bit. NOTE: Just commit your fixes to the same location, the pull request is updated automatically.

So FileSystem Testcases didn't' work until I changed the graddle can you add this to dependencies? As soon as I added it and imported, it started to recognize other libraries like @Rule and @tempFolder

[testCompileOnly 'junit:junit:4.12']

ZGreening commented on build.gradle:
```diff
@@ -12,6 +12,8 @@ dependencies {
     compile group: 'org.xerial', name: 'sqlite-jdbc', version: '3.27.2.1'
     testCompile group: 'org.junit.jupiter', name: 'junit-jupiter', version: '5.4.0'
+    testCompile("org.assertj:assertj-core:3.11.1")
+    testCompileOnly 'junit:junit:4.12'
```
I am not sure why junit Jupiter isn't giving you the dependencies. I guess we can merge this, but when I merge my updated one most of this will change anyway

ZGreening reviewed (approved):
Ok, I'd say it's good enough for now. Does Nathalie want to review? If not, I am fine for you to go ahead and merge
